addons/mail/controllers/attachment.py

```diff
@@ -67,50 +66,34 @@ def mail_attachment_upload(self, ufile, thread_id, thread_model, is_pending=Fals
             attachment = request.env["ir.attachment"].sudo().create(vals)
             attachment._post_add_create(**kwargs)
-            res = {"data": Store(attachment, extra_fields="access_token").get_result()}
+            res = {
+                "data": Store(
+                    attachment,
+                    extra_fields=request.env[
+                        "ir.attachment"
+                    ]._get_store_attachment_ownership_fields(),
+                ).get_result()
+            }
         except AccessError:
             res = {"error": _("You are not allowed to upload an attachment here.")}
         return request.make_json_response(res)
 
     @http.route("/mail/attachment/delete", methods=["POST"], type="jsonrpc", auth="public")
     @add_guest_to_context
-    def mail_attachment_delete(self, attachment_id, access_token=None, **kwargs):
+    def mail_attachment_delete(self, attachment_id, access_token=None):
         attachment = request.env["ir.attachment"].browse(int(attachment_id)).exists()
-        if not attachment:
+        if not attachment or not attachment._has_attachments_ownership([access_token]):
             request.env.user._bus_send("ir.attachment/delete", {"id": attachment_id})
-            return
-        attachment_message = request.env["mail.message"].sudo().search(
-            [("attachment_ids", "in", attachment.ids)], limit=1)
-        message = self._get_message_with_access(attachment_message.id, mode="create", **kwargs)
-        if not request.env.user.share:
-            # Check through standard access rights/rules for internal users.
-            attachment._delete_and_notify(message)
-            return
-        # For non-internal users 2 cases are supported:
-        #   - Either the attachment is linked to a message: verify the request is made by the author of the message (portal user or guest).
-        #   - Either a valid access token is given: also verify the message is pending (because unfortunately in portal a token is also provided to guest for viewing others' attachments).
-        # sudo: ir.attachment: access is validated below with membership of message or access token
-        attachment_sudo = attachment.sudo()
-        if message:
-            if not self._can_edit_message(message, **kwargs):
-                raise NotFound()
-        else:
-            if (
-                not access_token
-                or not attachment_sudo.access_token
-                or not consteq(access_token, attachment_sudo.access_token)
-            ):
-                raise NotFound()
-            if attachment_sudo.res_model != "mail.compose.message" or attachment_sudo.res_id != 0:
-                raise NotFound()
-        attachment_sudo._delete_and_notify(message)
+            raise NotFound()
+        message = request.env["mail.message"].search(
+            [("attachment_ids", "in", attachment.ids)], limit=1
+        )
```

Comment:
Should sudo search to avoid access rule filtering I think

Reply:
In this case, if the message is not accessible (not a logged in portal user), it's not possible to send a notification at all. So I don't think we need sudo here for now.

Reply:
Isn't it still possible for portal or internal users to access documents though token but have no rights on the message, while still being possible to receive bus notif for them?

This might only happen if there is a live chat on the same page though, as I don't think we start the bus on portal chatter otherwise.

But I think the code would be cleaner with the sudo here indeed. It was there before, I don't think we should be removing it in this PR.